[masonry] Optimize grid items with a span 1

stwrt opened the pull request:
#### f6831edd45a88907a4cf1d3d3d690338fb1db597
<pre>
[masonry] Optimize grid items with a span 1
<a href="https://bugs.webkit.org/show_bug.cgi?id=276385">https://bugs.webkit.org/show_bug.cgi?id=276385</a>
<a href="https://rdar.apple.com/problem/131407028">rdar://problem/131407028</a>

Reviewed by Sammy Gill.

General Case: grouping auto-placed items by span size -&gt; Nitems + Ntracks x Nspans
- compute aggregate items (Nitems)
- apply to tracks (Ntracks x Nspans)

This implements the algorithm for single track items. Items that span multiple tracks
and items that are &apos;fr&apos; will be handled in a follow up patch.

Runtime on microbenchark was reduced from 1.2 seconds to 38 ms.
Benchmark tested 1000 columns with &apos;auto&apos; and 10,000 grid items.
Tested on M1 Ultra.

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: <a href="https://commits.webkit.org/280919@main">https://commits.webkit.org/280919@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e39eac065df36632aa10e48668f3516ab2957cc4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58019 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37347 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10496 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61644 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8464 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60147 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8653 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47018 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6029 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60049 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35023 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50150 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27849 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31787 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7445 "Found 1 new test failure: imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-report-only-sends-reports-on-violation.https.sub.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7468 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53734 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7713 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63332 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1933 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7784 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54241 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1940 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50161 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54381 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12836 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1657 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33176 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34262 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35346 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34007 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/rendering/style/GridArea.h:
```diff
@@ -200,6 +199,8 @@ class GridSpan {
         m_endLine = std::max(GridPosition::min() + 1, std::min(endLine, GridPosition::max()));
     }
 
+private:
```
I needed to create a span, so I exposed the constructor. Maybe there is a better way of doing this.

sammygill replied:
I'm not 100% sure if this is ok (would need to think about it more) but for your usecase can you just use the static `translatedDefiniteGridSpan` function instead?

stwrt replied:
Yeah, we can probably just use that instead 😀

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -34,6 +34,7 @@
 #include "RenderGrid.h"
 #include "RenderStyleConstants.h"
 #include "StyleSelfAlignmentData.h"
+#include "platform/LayoutUnit.h"
```
Do we need this? I only ask since we were using `LayoutUnit`s before. If so we should remove the include should just be `LayoutUnit.h`

stwrt replied:
It does compile without it; removed 😀.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1479,10 +1505,18 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
 
 void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
 {
+    // WTF_ALWAYS_LOG("Entered resolveIntrinsicTrackSizes");
```
Stray comment/logging

stwrt replied:
Fixed

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1510,18 +1564,42 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
+    LayoutUnit singleTrackMinContentForChild;
+    LayoutUnit singleTrackMaxContentForChild;
+    LayoutUnit singleTrackMinSizeForChild;
+    
```
I feel like it would make sense to group all of these variables into a single struct since they all seem related. If we are only going to use these for masonry then maybe that should be included in the struct's name as well

stwrt replied:
That seems reasonable. I'll combine these all into a struct.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1510,18 +1564,42 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
 
         if (renderGrid()->isMasonry())
-            computeIndefiniteItems(masonryIndefiniteItems);
+            computeIndefiniteItems(singleTrackMasonryIndefiniteItems, masonryIndefiniteItems, singleTrackMinContentForChild, singleTrackMaxContentForChild, singleTrackMinSizeForChild);
```
Since all of these arguments are just out parameters I think we can refactor this function to just return a struct instead. If we refactor the fields that I mentioned above then we could just make it so that this function creates and returns that struct

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1510,18 +1564,42 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
+     * In case we hit an 'fr' track we will fallback to computing all indefinte items in the track.
+     *
+     * TODO: Add support for multi-track items.
+     * TDOO: Add support for flex items.
```
We generally represent these as `FIXME`s

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1492,12 +1526,32 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
+                            singleTrackMaxContentForChild = maxContentForChild;
+
+                        if (singleTrackMinSizeForChild < minSizeForChild)
+                            singleTrackMinSizeForChild = minSizeForChild;
```
We can write these as `foo = std::max(foo, bar)`

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1492,12 +1526,32 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
+                if (GridPositionsResolver::resolveGridPositionsFromStyle(*m_renderGrid, *gridItem, m_direction).isIndefinite()) {
+                    auto spanLength = m_renderGrid->gridSpanForChild(*gridItem, m_direction).integerSpan();
+
+                    if (spanLength == 1 && !gridItem->isRenderGrid()) {
```
Is there a reason why we do not allow grid items which are grids? In the comment you mention we do not include subgrids in this optimization but this by itself will also exclude grids which are grid items. If we want to exclude subgrids here we will want to expand on this condition a bit

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1479,10 +1505,18 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+            CheckedPtr<RenderBox> gridItem = it.get();
+            accumulateIntrinsicSizes(gridItem.get());
+        }
+    }
```
This logic seems to kinda come out of nowhere and it took me a bit to understand what its intent was in terms of the optimization/patch. Correct me if I am wrong, but is this to handle the `In case we hit an 'fr' track we will fallback to computing all indefinite items in the track` portion of your comment? 

If so, can we find a way to restructure the code so that we do not even attempt this optimization at all if we find some sort of content that we currently don't support? I think that would really improve the readability of the code instead of trying to interleave logic that tries to put us back into a state as if we did not try to do any sort of optimization.

stwrt replied:
Exactly, if we hit an 'fr' track we can't apply the optimization.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -270,6 +271,30 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
     return infinity;
 }
 
+void GridTrackSizingAlgorithm::sizeTrackToFitNonSpanningItem(const GridSpan& span, LayoutUnit& singleTrackMinContentForChild, LayoutUnit& singleTrackMaxContentForChild, LayoutUnit& singleTrackMinSizeForChild, GridTrack& track)
```
I'm not convinced this is the best name for this function. While the logic inside does seem similar to the version it overloads at a glance, I think this is doing something much different. It's taking some `LayoutUnit` arguments which represent some sort of sizes and trying to fit the track to them.

sammygill replied:
Do we need to pass these `LayoutUnit`s by reference? They don't seem to be out parameters and I think we generally pass them by value since they are small

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -270,6 +270,28 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
     return infinity;
 }
 
+void GridTrackSizingAlgorithm::sizeTrackToFitSingleSpanMasonryGroup(const GridSpan& span, struct MasonryIndefiniteItems &masonryIndefiniteItems, GridTrack& track)
```
I thiiink we don't need the `struct` here and this could just be `MasonryIndefiniteItems& masonryIndefiniteItems`? Same thing should apply to all other times you define a variable/parameter outside of the `struct` definition

stwrt replied:
Updating to `subgridMasonryIndefiniteItems` to give a better description.

We are trying to get all the children of the gridItem, so it might be a bit confusing, since we already have `masonryIndefiniteItems` above.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -270,6 +270,28 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
 
+void GridTrackSizingAlgorithm::sizeTrackToFitSingleSpanMasonryGroup(const GridSpan& span, struct MasonryIndefiniteItems &masonryIndefiniteItems, GridTrack& track)
+{
+    unsigned trackPosition = span.startLine();
```
I think this should be auto to follow the general style of the rest of the code

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1475,9 +1497,25 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
-        CheckedPtr<RenderBox> gridItem = it.get();
-        accumulateIntrinsicSizes(gridItem.get());
+    if (m_renderGrid->isMasonry()) {
+        for (auto it = masonryIndefiniteItems.indefiniteItems.begin(), end = masonryIndefiniteItems.indefiniteItems.end(); it != end; ++it) {
```
Can we use a modern C++ for loop here? If not we should be able to rewrite this as `auto it = masonryIndefiniteItems.indefiniteItems.begin(), it != masonryIndefiniteItems.indefiniteItems.end(); ++it` to be a bit more succinct

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1492,12 +1530,34 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
+
+                    bool isSubgrid = false;
+                    if (gridItem->isRenderGrid()) {
+                        CheckedPtr inner = dynamicDowncast<RenderGrid>(gridItem.get());
```
We don't need this `CheckedPtr` here since the pointer is already covered by the `CheckedPtr<RenderBox> gridItem = iterator.nextGridItem()` above. Also since we already know this renderer is a `RenderGrid` we could probably just do a `downcast`

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -270,6 +270,28 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
+    if (trackSize.hasMinContentMaxTrackBreadth())
+        track.setGrowthLimit(std::max(track.growthLimit(), masonryIndefiniteItems.largestMinContentSizeForSingleTrackItems));
+    else if (trackSize.hasMaxContentOrAutoMaxTrackBreadth()) {
+        LayoutUnit growthLimit = masonryIndefiniteItems.largestMaxContentSizeForSingleTrackItems;
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1444,12 +1466,12 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
 
             GridIterator childIterator = GridIterator::createForSubgrid(*inner, iterator, subgridSpan);
-            SingleThreadWeakListHashSet<RenderBox> childMasonryIndefiniteItems;
+            struct MasonryIndefiniteItems childMasonryIndefiniteItems;
```
What is child adding here? Since we're already touching this variable should this just be `masonryIndefiniteItems`?

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1475,9 +1497,25 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+        if (!(track.cachedTrackSize().minTrackBreadth().isFlex() || track.cachedTrackSize().maxTrackBreadth().isFlex()))
+            sizeTrackToFitSingleSpanMasonryGroup(GridSpan::translatedDefiniteGridSpan(trackIndex, trackIndex + 1), masonryIndefiniteItems, track);
+        else {
+            for (auto it = masonryIndefiniteItems.singleTrackIndefiniteItems.begin(), end = masonryIndefiniteItems.singleTrackIndefiniteItems.end(); it != end; ++it) {
```
Same comment about the for loop as above

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1492,12 +1530,34 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
+                    if (spanLength == 1 && !isSubgrid) {
+                        LayoutUnit minContentForChild = m_strategy->minContentForChild(*gridItem);
+                        LayoutUnit maxContentForChild = m_strategy->maxContentForChild(*gridItem);
+                        LayoutUnit minSizeForChild = m_strategy->minSizeForChild(*gridItem);
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -34,6 +34,7 @@
 #include "RenderGrid.h"
 #include "RenderStyleConstants.h"
 #include "StyleSelfAlignmentData.h"
+#include "rendering/RenderObject.h"
```
This `#include` is trying to sneak in (pretty sure it's unnecessary)

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -270,6 +271,28 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
     return infinity;
 }
 
+void GridTrackSizingAlgorithm::sizeTrackToFitSingleSpanMasonryGroup(const GridSpan& span, MasonryIndefiniteItems &masonryIndefiniteItems, GridTrack& track)
```
There is an extra space here and this should be `MasonryIndefiniteItems& masonryIndefiniteItems`

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1492,12 +1527,31 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
     };
 
-    auto computeIndefiniteItems = [&](SingleThreadWeakListHashSet<RenderBox>& indefiniteItems) {
+    auto computeIndefiniteItems = [&](MasonryIndefiniteItems &masonryIndefiniteItems) {
```
Extra space here too should be `MasonryIndefiniteItems& masonryIndefiniteItems`

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
```diff
@@ -202,7 +227,7 @@ class GridTrackSizingAlgorithm final {
     void stretchAutoTracks();
 
-    void accumulateIntrinsicSizesForTrack(GridTrack&, unsigned trackIndex, GridIterator&, Vector<GridItemWithSpan>& itemsSortedByIncreasingSpan, Vector<GridItemWithSpan>& itemsCrossingFlexibleTracks, SingleThreadWeakHashSet<RenderBox>& itemsSet, SingleThreadWeakListHashSet<RenderBox>& masonryIndefiniteItems, LayoutUnit currentAccumulatedMbp);
+    void accumulateIntrinsicSizesForTrack(GridTrack&, unsigned trackIndex, GridIterator&, Vector<GridItemWithSpan>& itemsSortedByIncreasingSpan, Vector<GridItemWithSpan>& itemsCrossingFlexibleTracks, SingleThreadWeakHashSet<RenderBox>& itemsSet, MasonryIndefiniteItems &, LayoutUnit currentAccumulatedMbp);
```
Extra space here too

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1475,9 +1497,21 @@ void GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack(GridTrack& track
+        * If we are in a non flex track we can then skip having to evaluate every item.
+        * 
+        * FIXME: Add support for optimizing the flex track scenario.
+        */
```
While I did not see it in our style guide, it looks like the rest of the code uses `//` even for comments over multiple lines

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
```diff
@@ -162,6 +162,29 @@ class GridTrackSizingAlgorithm final {
+         *
+         * FIXME: Add support for multi-track items.
+         * FIXME: Add support for flex items.
+         */
```
Same remark/question as above regarding comments?

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1492,12 +1526,31 @@ void GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes()
+
+                    bool isSubgrid = false;
+                    if (gridItem->isRenderGrid() && downcast<RenderGrid>(gridItem.get())->isSubgridInParentDirection(iterator.direction()))
+                        isSubgrid = true;
```
You might be able to refactor this into a single line if you think it doesn't harm readability

stwrt committed:
[masonry] Optimize grid items with a span 1
https://bugs.webkit.org/show_bug.cgi?id=276385
rdar://problem/131407028

Reviewed by Sammy Gill.

General Case: grouping auto-placed items by span size -> Nitems + Ntracks x Nspans
- compute aggregate items (Nitems)
- apply to tracks (Ntracks x Nspans)

This implements the algorithm for single track items. Items that span multiple tracks
and items that are 'fr' will be handled in a follow up patch.

Runtime on microbenchark was reduced from 1.2 seconds to 38 ms.
Benchmark tested 1000 columns with 'auto' and 10,000 grid items.
Tested on M1 Ultra.

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::accumulateIntrinsicSizesForTrack):
(WebCore::GridTrackSizingAlgorithm::resolveIntrinsicTrackSizes):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Canonical link: https://commits.webkit.org/280919@main
